### DISCUSSION TITLE
[BUG] InResponseTo not valid for all requests

### DISCUSSION THREAD

**Discussion by innovatechs** [bergie/passport-saml#683] [General]
Hi everyone

We are getting the InResponseTo not valid error for each request since the CacheProvider is creating a new instance for each call (our  assumption): getting empty keys when we call the get method.

**Expected behavior**

The previous versions were using one instance of the CacheProvider (from constructor) and keys are managed correctly (save, get and delete).

**Environment**

Node.js version: 14
passport-saml version: 3.2.1

**Comment by srd90** [comment]
You didn't care to define what was your previous version.

Lets assume your previous version was https://github.com/sign-in-canada/passport-saml/tree/275f15488823407043208677caf3dea8fee9903a (current HEAD of `sign-in-canada/passport-saml` repo's `master` branch which seems to be based on `node-saml/passport-saml 2.x`).

And lets assume that you are merging `passport-saml 3.2.1` (based on content of https://github.com/sign-in-canada/passport-saml/tree/b25305b50e0f69690d788ef48a1145d5bdab1fba which is current HEAD of `sign-in-canada/passport-saml` repo's `3.x` branch).

Since you have modified (based on this commit https://github.com/sign-in-canada/passport-saml/commit/e4fe1fc25a9a694cc980e6045824b6bde7dd1b7a ... which btw. doesn't make any sense) also multisamlstrategy (which is pretty much the only place which recreates `SAML` object each and every time) you might have encounter this issue in the context of using multisamlstrategy.

`node-saml/passport-saml 2.x`'s multisamlstrategy created implicitly shared InMemoryCacheProvider if `cacheProvider` option was not defined:
https://github.com/node-saml/passport-saml/blob/0d983aa7806c4529efef75c7534cfdf9b7bd04a3/src/passport-saml/multiSamlStrategy.ts#L19-L22

`node-saml/passport-saml 3.x` doesn't do that anymore (related comments: https://github.com/node-saml/passport-saml/pull/548#pullrequestreview-807397405 and https://github.com/node-saml/passport-saml/issues/334#issuecomment-971653795).

**Wild guess**:
1. You have typo in your `cacheProvider` configuration option name and
2. when multisamlstrategy creates SAML object its constructor creates new InMemoryCacheProvider due missing `cacheProvider` option during e.g. each `authenticate` invocation. 

I.e. code flow would/could be like this (links point to current HEAD of `node-saml/passport-saml 3.x`):
https://github.com/node-saml/passport-saml/blob/6ba76ba3a015fea96a2dd38f661a6c1f85bc44a1/src/passport-saml/multiSamlStrategy.ts#L35-L36
https://github.com/node-saml/passport-saml/blob/6ba76ba3a015fea96a2dd38f661a6c1f85bc44a1/src/passport-saml/multiSamlStrategy.ts#L41
https://github.com/node-saml/passport-saml/blob/6ba76ba3a015fea96a2dd38f661a6c1f85bc44a1/src/node-saml/saml.ts#L160

**Comment by cjbarth** [comment]
I'm in agreement with @srd90 here. I don't think we have enough information to say what exactly the problem is. Moving to discussion.
